### PR TITLE
Global exception handler

### PR DESCRIPTION
**Pull request by Parashchuk**

### CODE REVIEW

**Review by SalnikovMarkiian — APPROVED** [review]
LGTM

File: pom.xml [hunk 1]
```diff
@@ -71,6 +71,11 @@
             <artifactId>mapstruct</artifactId>
             <version>${mapstruct.version}</version>
         </dependency>
+
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
+            <artifactId>hibernate-validator</artifactId>
```
Comment (NazarSmith228) [inline pom.xml]: it is better to use `spring-boot-starter-validation`

File: src/main/java/bookstore/controller/BookController.java [hunk 2]
```diff
@@ -36,13 +37,14 @@ public BookDto getBookById(@PathVariable Long id) {
...
     @ResponseStatus(HttpStatus.CREATED)
-    public BookDto createBook(@RequestBody CreateBookRequestDto book) {
+    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto book) {
         return bookService.save(book);
     }
 
     @PutMapping("/{id}")
     @ResponseStatus(HttpStatus.OK)
-    public BookDto updateBookById(@PathVariable Long id, @RequestBody CreateBookRequestDto book) {
+    public BookDto updateBookById(@PathVariable Long id,
```
Comment (NazarSmith228) [inline src/main/java/bookstore/controller/BookController.java]: you can also validate `@PathVariable` like:
```
@PathVariable @Positive Long id
```

in order to do so - annotate you controller with `@Validated`

File: src/main/java/bookstore/dto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -1,12 +1,31 @@
...
 
+import bookstore.validation.Isbn;
+import jakarta.validation.constraints.DecimalMin;
+import jakarta.validation.constraints.Max;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Size;
 import java.math.BigDecimal;
 
 public record CreateBookRequestDto(
+        @NotNull
```
Comment (NazarSmith228) [inline src/main/java/bookstore/dto/CreateBookRequestDto.java]: it is better to use `@NotBlank`

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            HttpHeaders headers,
+            HttpStatusCode status,
+            WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        List<String> errors = ex.getBindingResult().getAllErrors().stream()
+                .map(this::getErrorMessage)
+                .toList();
+
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
```
Comment (NazarSmith228) [inline src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]: it makes no sense to store status in body

status is obvious from the response itself

File: src/main/java/bookstore/validation/IsbnValidator.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package bookstore.validation;
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class IsbnValidator implements ConstraintValidator<Isbn, String> {
+
+    @Override
+    public boolean isValid(String s, ConstraintValidatorContext constraintValidatorContext) {
+        return s != null && s.length() == 13;
```
Comment (NazarSmith228) [inline src/main/java/bookstore/validation/IsbnValidator.java]: such validation wil consider such string: `             ` as valid :)

it is better to use some regex

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java (hunk 4 above)
Comment (Parashchuk) [reply src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]: Doesn't a client-side developer need a status code to create custom logic?

File: src/main/java/bookstore/validation/IsbnValidator.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package bookstore.validation;
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class IsbnValidator implements ConstraintValidator<Isbn, String> {
+    private static final String ISBN_REGEX = "\\d{13}";
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/validation/IsbnValidator.java:7]: You can use Pattern here

File: src/main/java/bookstore/validation/IsbnValidator.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+    @Override
+    public boolean isValid(String s, ConstraintValidatorContext constraintValidatorContext) {
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/validation/IsbnValidator.java]: s - bad naming

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            HttpHeaders headers,
+            HttpStatusCode status,
+            WebRequest request
+    ) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        List<String> errors = ex.getBindingResult().getAllErrors().stream()
+                .map(this::getErrorMessage)
+                .toList();
+
+        body.put("status", HttpStatus.BAD_REQUEST.value());
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]: Don't need to store status and time here

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    @ExceptionHandler(EntityNotFoundException.class)
+    public ResponseEntity<Map<String, Object>> handleEntityNotFoundException(
+            EntityNotFoundException ex
+    ) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.NOT_FOUND.value());
+        body.put("error", HttpStatus.NOT_FOUND);
+        body.put("message", ex.getMessage());
+        return new ResponseEntity<>(body, new HttpHeaders(), HttpStatus.NOT_FOUND);
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java:43]: You can easily return new ResponseEntity with status and ex.message

File: src/main/java/bookstore/exception/CustomEntityNotFoundResponse.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package bookstore.exception;
+
+public class CustomEntityNotFoundResponse {
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/exception/CustomEntityNotFoundResponse.java]: Where do you use it? Can't find

File: src/main/java/bookstore/controller/BookController.java [hunk 11]
```diff
@@ -36,18 +40,21 @@ public BookDto getBookById(@PathVariable Long id) {
...
+            @PathVariable @Positive Long id,
+            @Valid @RequestBody CreateBookRequestDto book
+    ) {
         return bookService.updateById(id, book);
     }
 
     @DeleteMapping("/{id}")
-    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
     public void deleteBookById(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/BookController.java]:
```suggestion
    public void deleteBookById(@PathVariable @Positive Long id) {
```

File: src/main/java/bookstore/controller/BookController.java [hunk 12]
```diff
@@ -36,18 +40,21 @@ public BookDto getBookById(@PathVariable Long id) {
...
         bookService.deleteById(id);
     }
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/BookController.java:60]: What about adding validation to public BookDto getBookById(@PathVariable @Positive Long id) {

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**